### PR TITLE
Fix APIMANAGER-4369

### PR DESCRIPTION
**Pull request by praminda**

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 1]
```diff
@@ -1066,6 +1066,42 @@ public API getAPI(APIIdentifier identifier,APIIdentifier oldIdentifier, String o
...
+        Set<Tier> tiers = new TreeSet<Tier>(new TierNameComparator());
+
+        Map<String, Tier> tierMap;
+        if (tenantId == MultitenantConstants.INVALID_TENANT_ID) {
+            tierMap = APIUtil.getAllTiers();
+        } else {
+            PrivilegedCarbonContext.startTenantFlow();
+            PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantId(tenantId, true);
+            tierMap = APIUtil.getAllTiers(tenantId);
+            PrivilegedCarbonContext.endTenantFlow();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: startTenantFlow need to be in try block and endTenantFlow should done in finally block

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 2]
```diff
@@ -1066,6 +1066,42 @@ public API getAPI(APIIdentifier identifier,APIIdentifier oldIdentifier, String o
...
+
+        Map<String, Tier> tierMap;
+        int requestedTenantId = PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
+        if (requestedTenantId == 0 || requestedTenantId == MultitenantConstants.INVALID_TENANT_ID) {
+            tierMap = APIUtil.getAllTiers();
+        } else {
+            tierMap = APIUtil.getAllTiers(requestedTenantId);
+        }
+        tiers.addAll(tierMap.values());
+        PrivilegedCarbonContext.endTenantFlow();
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: startTenantFlow need to be in try block and endTenantFlow should done in finally block

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 3]
```diff
@@ -1776,6 +1835,29 @@ public static APIStore getExternalAPIStore(String apiStoreName, int tenantId) th
...
+     * @throws RegistryException when registry action fails
+     * @throws XMLStreamException when xml parsing fails
+     * @throws APIManagementException when fails to retrieve tier attributes
+     */
+    private static Map<String, Tier> getTiers(Registry registry,String tierLocation)
+            throws RegistryException, XMLStreamException, APIManagementException {
+        Map<String, Tier> tiers = getAllTiers(registry, tierLocation);
+        try {
+            tiers.remove(APIConstants.UNAUTHENTICATED_TIER);
+        } catch (Exception e) {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: any reason for catching generic exception?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java [hunk 4]
```diff
@@ -1776,6 +1835,29 @@ public static APIStore getExternalAPIStore(String apiStoreName, int tenantId) th
...
+     *
+     * @param registry registry to access tiers config
+     * @param tierLocation registry location of tiers config
+     * @return map containing available tiers
+     * @throws RegistryException when registry action fails
+     * @throws XMLStreamException when xml parsing fails
+     * @throws APIManagementException when fails to retrieve tier attributes
+     */
+    private static Map<String, Tier> getTiers(Registry registry,String tierLocation)
+            throws RegistryException, XMLStreamException, APIManagementException {
```
Comment (harsha89) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: and method throws multiple exceptions even catching a generic exception?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java (hunk 1 above)
Comment (thilinicooray) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: Best practice would be to use following approach

try {
            if (tenantDomain != null && !MultitenantConstants.SUPER_TENANT_DOMAIN_NAME.equals(tenantDomain)) {
                PrivilegedCarbonContext.startTenantFlow();
                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
                isTenantFlowStarted = true;
            }
            // your logic goes here.

        } finally {
            if (isTenantFlowStarted) {
                PrivilegedCarbonContext.endTenantFlow();
            }
        }

Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: Fixed in #1672

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java (hunk 2 above)
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java]: Fixed in #1672

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java (hunk 3 above)
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: Fixed in #1672

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java (hunk 4 above)
Comment (praminda) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/utils/APIUtil.java]: Fixed in #1672

**Pull request merged**